debian/grub/25_linux_acrn

```diff
@@ -97,117 +121,151 @@ fi
 
 case x"$GRUB_FS" in
     xbtrfs)
-	rootsubvol="`make_system_path_relative_to_its_root /`"
-	rootsubvol="${rootsubvol#/}"
-	if [ "x${rootsubvol}" != x ]; then
-	    GRUB_CMDLINE_LINUX="rootflags=subvol=${rootsubvol} ${GRUB_CMDLINE_LINUX}"
-	fi;;
+        rootsubvol="$(make_system_path_relative_to_its_root /)"
+        rootsubvol="${rootsubvol#/}"
+        if [ "x${rootsubvol}" != x ]; then
+            GRUB_CMDLINE_LINUX="rootflags=subvol=${rootsubvol} ${GRUB_CMDLINE_LINUX}"
+        fi;;
     xzfs)
-	rpool=`${grub_probe} --device ${GRUB_DEVICE} --target=fs_label 2>/dev/null || true`
-	bootfs="`make_system_path_relative_to_its_root / | sed -e "s,@$,,"`"
-	LINUX_ROOT_DEVICE="ZFS=${rpool}${bootfs%/}"
-	;;
+        rpool=$(${grub_probe} --device ${GRUB_DEVICE} --target=fs_label 2>/dev/null || true)
+        bootfs="$(make_system_path_relative_to_its_root / | sed -e "s,@$,,")"
+        LINUX_ROOT_DEVICE="ZFS=${rpool}${bootfs%/}"
+        ;;
 esac
 
-title_correction_code=
+add_prelaunched_vms()
+{
+    # eventually also load data for pre-launched VMs
+    for id in ${ACRN_VM_IDS}; do
+        # we just care for pre-launched VMs
+        [ "$(eval echo \${ACRN_LOAD_ORDER_VM${id}})" != "PRE_LAUNCHED_VM" ] && continue
+
+        if [ -z "$(eval echo \${GRUB_ACRN_KERNEL_VM${id}})" ]; then
```

Comment:
VM IDs can be quite volatile and changes from scenario to scenario. I would recommend using the module tags instead.

As an example, a user can set the tag of a pre-launched VM kernel to `Linux_bzImage` and define `ACRN_KERN_MOD_Linux_bzImage` in acrn.cfg to specify where that kernel is in the filesystem. That allows users to simply use the same tag for the same kernel in their scenarios and not to worry about the ordering of the pre-launched VMs.

Reply:
This is a good suggestion. Using the the module tags instead of VM ids is way better. I'll change that!